<a href="https://colab.research.google.com/github/mbparsons/code/blob/master/Basic_AD_Analyses.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

In [0]:
import pandas as pd
import numpy as np
import seaborn as sns

import statsmodels.api as sm

import sklearn as sk
from statsmodels.stats.outliers_influence import variance_inflation_factor
from sklearn.linear_model import LassoCV, LarsCV, LassoLarsCV
from sklearn.metrics import mean_squared_error, r2_score, mean_absolute_error, accuracy_score

In [0]:
from google.colab import drive
drive.mount('/content/drive')

Go to this URL in a browser: https://accounts.google.com/o/oauth2/auth?client_id=947318989803-6bn6qk8qdgf4n4g3pfee6491hc0brc4i.apps.googleusercontent.com&redirect_uri=urn%3Aietf%3Awg%3Aoauth%3A2.0%3Aoob&scope=email%20https%3A%2F%2Fwww.googleapis.com%2Fauth%2Fdocs.test%20https%3A%2F%2Fwww.googleapis.com%2Fauth%2Fdrive%20https%3A%2F%2Fwww.googleapis.com%2Fauth%2Fdrive.photos.readonly%20https%3A%2F%2Fwww.googleapis.com%2Fauth%2Fpeopleapi.readonly&response_type=code

Enter your authorization code:
··········
Mounted at /content/drive


In [0]:
pony = 'drive/My Drive/Colab Notebooks/Travel Pony Facebook.csv'
df = pd.read_csv(pony)

In [0]:
df['Cost per Impression'] = df['Amount Spent (USD)']/df['Impressions']

In [0]:
df

Start Date  ... Cost per Impression
0     2013-12-02  ...            0.000466
1     2013-12-01  ...            0.000652
2     2013-11-30  ...            0.000448
3     2013-11-29  ...            0.000452
4     2013-11-22  ...            0.002464
...          ...  ...                 ...
3700  2013-06-11  ...            0.003512
3701  2013-06-10  ...            0.002936
3702  2013-06-09  ...            0.008919
3703  2013-06-08  ...            0.006643
3704  2013-06-07  ...            0.000439

[3705 rows x 21 columns]

In [0]:
!pip install arrow

     |████████████████████████████████| 51kB 1.7MB/s 


In [0]:
import arrow

In [0]:
sd = df['Start Date']
sd

0       2013-12-02
1       2013-12-01
2       2013-11-30
3       2013-11-29
4       2013-11-22
           ...    
3700    2013-06-11
3701    2013-06-10
3702    2013-06-09
3703    2013-06-08
3704    2013-06-07
Name: Start Date, Length: 3705, dtype: object

In [0]:
utc = arrow.utcnow()

In [0]:
now = arrow.utcnow()
date_time_zone = now.format('YYYY-MM-DD HH:mm:ss ZZ')
print("Date and time and zone: {0}".format(date_time_zone))

Date and time and zone: 2019-11-14 06:23:02 +00:00


In [0]:
a = arrow.get(sd[0])
a.weekday()
a.format('dddd')

<Arrow [2013-12-02T00:00:00+00:00]>

In [0]:
w= []
for i in range(len(sd)):
  d = arrow.get(sd[i])
  w.append(d.format('dddd'))

#print(d.weekday())
#print(d.format('dddd'))

In [0]:
w

['Monday',
 'Sunday',
 'Saturday',
 'Friday',
 'Friday',
 'Friday',
 'Thursday',
 'Thursday',
 'Wednesday',
 'Wednesday',
 'Tuesday',
 'Tuesday',
 'Monday',
 'Sunday',
 'Saturday',
 'Friday',
 'Friday',
 'Thursday',
 'Thursday',
 'Thursday',
 'Thursday',
 'Thursday',
 'Wednesday',
 'Wednesday',
 'Wednesday',
 'Wednesday',
 'Tuesday',
 'Tuesday',
 'Tuesday',
 'Tuesday',
 'Monday',
 'Monday',
 'Monday',
 'Monday',
 'Monday',
 'Sunday',
 'Sunday',
 'Sunday',
 'Sunday',
 'Sunday',
 'Sunday',
 'Saturday',
 'Saturday',
 'Saturday',
 'Saturday',
 'Saturday',
 'Saturday',
 'Friday',
 'Friday',
 'Friday',
 'Friday',
 'Friday',
 'Friday',
 'Thursday',
 'Thursday',
 'Wednesday',
 'Wednesday',
 'Wednesday',
 'Wednesday',
 'Wednesday',
 'Wednesday',
 'Wednesday',
 'Wednesday',
 'Wednesday',
 'Wednesday',
 'Wednesday',
 'Wednesday',
 'Wednesday',
 'Wednesday',
 'Wednesday',
 'Wednesday',
 'Wednesday',
 'Wednesday',
 'Tuesday',
 'Tuesday',
 'Tuesday',
 'Tuesday',
 'Tuesday',
 'Tuesday',
 'Tuesday',
 

In [0]:
df['Weekday'] = w 

In [0]:
df

Start Date    End Date  ... Cost per Impression   Weekday
0     2013-12-02  2013-12-02  ...            0.000466    Monday
1     2013-12-01  2013-12-01  ...            0.000652    Sunday
2     2013-11-30  2013-11-30  ...            0.000448  Saturday
3     2013-11-29  2013-11-29  ...            0.000452    Friday
4     2013-11-22  2013-11-22  ...            0.002464    Friday
...          ...         ...  ...                 ...       ...
3700  2013-06-11  2013-06-11  ...            0.003512   Tuesday
3701  2013-06-10  2013-06-10  ...            0.002936    Monday
3702  2013-06-09  2013-06-09  ...            0.008919    Sunday
3703  2013-06-08  2013-06-08  ...            0.006643  Saturday
3704  2013-06-07  2013-06-07  ...            0.000439    Friday

[3705 rows x 22 columns]

In [0]:
pd.pivot_table(df,index=["Weekday"],values=["Cost per Impression"])

Cost per Impression
Weekday                       
Friday                0.004097
Monday                0.002910
Saturday              0.002629
Sunday                0.003688
Thursday              0.003398
Tuesday               0.003040
Wednesday             0.002981

In [0]:
weekdayMean = pd.pivot_table(df,index=["Weekday"],values=["Cost per Impression"])

In [0]:
weekdayMean.sort_values(by = 'Cost per Impression', inplace=True)
weekdayMean

Cost per Impression
Weekday                       
Saturday              0.002629
Monday                0.002910
Wednesday             0.002981
Tuesday               0.003040
Thursday              0.003398
Sunday                0.003688
Friday                0.004097

1. When considering 'cost per impression', what day of the week works best? (What day is it cheapest to generate impressions) 
What day works worst? (What day is most expensive)


Saturday is the best day of the week because average cost per impression is the least (0.002629). Friday works the worst because the average cost per impression is the most expensive (0.004097).

2. Compute the correlation between Amount Spent and the following variables:
- Reach
- Frequency 
- Unique Clicks
- Page Likes

Which correlation is the strongest? What does that mean practically? (respond in a tweet or less)

The correlation between Amount Spent and Unique Clicks is the strongest. Practically that means paying more for advertisements generates more unique clicks.

In [0]:
print( df[['Amount Spent (USD)','Reach']].corr())
print( df[['Amount Spent (USD)','Frequency']].corr())
print( df[['Amount Spent (USD)','Unique Clicks']].corr())
print( df[['Amount Spent (USD)','Page Likes']].corr())

                    Amount Spent (USD)     Reach
Amount Spent (USD)            1.000000  0.703124
Reach                         0.703124  1.000000
                    Amount Spent (USD)  Frequency
Amount Spent (USD)            1.000000   0.130201
Frequency                     0.130201   1.000000
                    Amount Spent (USD)  Unique Clicks
Amount Spent (USD)            1.000000       0.882993
Unique Clicks                 0.882993       1.000000
                    Amount Spent (USD)  Page Likes
Amount Spent (USD)            1.000000    0.757612
Page Likes                    0.757612    1.000000


3. Perform a simple multiple regression analysis where Unique Clicks is the dependent variable and Reach and Frequency are the independent (predictor) variables.

What variable most strongly predicts unique clicks? What does that mean practically? (respond in a tweet or less)

Reach most strongly predicts unique clicks because Reach has a much higher t-statistic than Frequency. The more people you reach the more likely you are to have unique people that click on an advertisement.



In [0]:
y = df['Unique Clicks']
x = df[['Reach', 'Frequency']]

model = sm.OLS(y, x).fit()
predictions = model.predict(x) # make the predictions by the model

# Print out the statistics
model.summary()

<class 'statsmodels.iolib.summary.Summary'>
"""
                                 OLS Regression Results                                
=======================================================================================
Dep. Variable:          Unique Clicks   R-squared (uncentered):                   0.557
Model:                            OLS   Adj. R-squared (uncentered):              0.556
Method:                 Least Squares   F-statistic:                              2325.
Date:                Thu, 14 Nov 2019   Prob (F-statistic):                        0.00
Time:                        07:48:19   Log-Likelihood:                         -15973.
No. Observations:                3705   AIC:                                  3.195e+04
Df Residuals:                    3703   BIC:                                  3.196e+04
Df Model:                           2                                                  
Covariance Type:            nonrobust                                                  
==============================================================================
                 coef    std err          t      P>|t|      [0.025      0.975]
------------------------------------------------------------------------------
Reach          0.0019   3.12e-05     62.490      0.000       0.002       0.002
Frequency      3.6139      0.298     12.109      0.000       3.029       4.199
==============================================================================
Omnibus:                     5107.616   Durbin-Watson:                   0.803
Prob(Omnibus):                  0.000   Jarque-Bera (JB):          5128392.747
Skew:                           7.331   Prob(JB):                         0.00
Kurtosis:                     184.674   Cond. No.                     9.84e+03
==============================================================================

Warnings:
[1] Standard Errors assume that the covariance matrix of the errors is correctly specified.
[2] The condition number is large, 9.84e+03. This might indicate that there are
strong multicollinearity or other numerical problems.
"""